text/3641-export-function-ordinals.md

```diff
@@ -0,0 +1,145 @@
+
+By creating an attribute for specifying function ordinals, we can choose the ordinal position for the functions where it matters, and let Rust choose the ordinal for any other functions where ordinal position is not important.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Ordinals
+
+Function Ordinals refer to the position of an exported function in a Dynamically Linked Library (DLL). When accessing functions by name, this is not important. However some applications access functions based on their position (ordinal), rather than their name. The Microsoft documentation for this concept is available [here.](https://learn.microsoft.com/en-us/cpp/build/exporting-functions-from-a-dll-by-ordinal-rather-than-by-name)
+
+## Usage
+
+You can specify the ordinality of an exported function using the `export_ordinal` attribute on it.
+
+```rs
+#[export_ordinal(1)]
+pub extern "C" fn hello() {
+    println!("Hello, World!");
+}
+```
+
+This example will export `hello` as ordinal 1, and when a program tries to call ordinal 1 in your DLL, it will be executed.
+
+## Behaviour
+
+If other software expects your function to be a specific ordinal, you should be very careful when changing the ordinal or removing the `export_ordinal` attribute, as it could lead to the wrong function being called (or not found at all).
+
+If `export_ordinal` isn't provided, an unused ordinal will be assigned during compilation.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`export_ordinal` is a new attribute for functions which has a signature similar to the following:
+
+```rs
+#[export_ordinal(n)]
+```
+
+`n` must be:
+
+1. A positive integer >= 1
+2. Unique across the entire program.
+   - An error should be thrown if the same ordinal is provided in multiple places.
+
+The attribute should only affect windows targets, as ordinals are not a feature of shared libraries on other targets.
+
+The attribute must be placed above an exported function like so:
+
+```rs
+#[no_mangle]
+#[export_ordinal(1)]
+pub fn hello() {}
+
+// Also works with extern and unsafe functions
+
+#[no_mangle]
+#[export_ordinal(2)]
+pub unsafe extern "C" fn world() {}
+```
+
+This should also work with exported bindings, such as:
+
+```rs
+#[link(name = "external_library")]
+extern "C" {
+    #[export_ordinal(2)]
+    pub fn external_function();
+}
+```
+
+If `#[export_ordinal(n)]` is not provided above an exported function, the current behaviour should apply, where an unused ordinal is provided sequentially. For example, if ordinals `1` and `3` are specified with the `export_ordinal` attribute, and there are two other exported functions without specified ordinals, they should use ordinals `2` and `4`.
```

Comment:
This is also listed as an unresolved question below. Maybe it's better to leave it unresolved, since I think it sounds difficult to implement in a strict sense.

Reply:
Thanks, this was an oversight - I've removed this text and kept it in the unresolved questions